### PR TITLE
Consolidate Concepts

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR:
  * Adds the Introduction section.
  * Adds Development Process page under the Introduction.
  * Transforms the Concepts section into one page under the Introduction section.
  * Moves the Architecture page under the Introduction section.
  * Changes the Documentation home page to reflect new structure.

### CODE REVIEW

File: docs/introduction/concepts.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+## Value Objects
+
+Value Object describe things in a domain model and do not have identity. 
+Value Objects are also immutable. Some examples are:
+  * `PhoneNumber`
+  * `EmailAddress`
+  * `BarCode`
+  
+In Spine, Value Objects are defined as Protobuf messages.
```
Comment (armiol) [inline docs/introduction/concepts.md]: Somewhere on this page, we should say `Protocol Buffers (a.k.a. Protobuf)` when we first mention it.

File: docs/introduction/concepts.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+## Entities
+
+Entities are main building blocks of a domain model. They have unique identity and modify their
```
Comment (armiol) [inline docs/introduction/concepts.md]: "...are the main..."
"They have a unique identity..."

File: docs/introduction/concepts.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+  * `PhoneNumber`
+  * `EmailAddress`
+  * `BarCode`
+  
+In Spine, Value Objects are defined as Protobuf messages.
+
+## Entities
+
+Entities are main building blocks of a domain model. They have unique identity and modify their 
+state during the lifecycle.
```
Comment (armiol) [inline docs/introduction/concepts.md]: "during their lifespan". 

"Lifecycle" is something different to me.

File: docs/introduction/concepts.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+### Identifier
+ 
+The framework supports the following types of identifiers:
+
+* `Integer`,
+* `Long`,
+* `String`,
+* A generated Java class implementing the `Message`.
```
Comment (armiol) [inline docs/introduction/concepts.md]: Maybe we should put this first?

File: docs/introduction/concepts.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+Aggregate is the main building block of a business model. 
+From the application point of view it consists of the following:
+1. Commands which arrive to it. 
+2. Events which appear in response to these commands. 
+3. How these events influence the state of an aggregate.
+
+[Aggregates](http://martinfowler.com/bliki/DDD_Aggregate.html) guarantee consistency of data
+modifications in response to commands they receive. Aggregate is the most common case of
+Command Handler. It modifies its state and produces one or more events in response to a command.
```
Comment (armiol) [inline docs/introduction/concepts.md]: I would rephrase to 
```
In response to a command, it produces one or more events modifying own state.
```

File: docs/introduction/concepts.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+These events are used later to restore the state of the aggregate.
+
+In Spine, aggregates are defined as Java classes, and their states are defined as Protobuf messages.
+
+### Process Manager
+
+Process Manager is an independent component that reacts to domain events in a cross-aggregate
```
Comment (armiol) [inline docs/introduction/concepts.md]: I would rephrase.

```
Process Manager is an independent component that manages the cross-aggregate business flows. 
It serves as a mediator by remembering the state of the flow and choosing the next step 
based on the intermediate results.

To do so, a Process Manager can be both [Command Handler](#command-handler) 
and [Event Reactor](#event-reactor). Also, it can emit Commands to other Aggregates 
and Process Managers.
```

File: docs/introduction/concepts.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+Projection is an [Event Subscriber](#event-subscriber) which transforms multiple events data into
+a structural representation. Projections are the main building blocks of the Query side of
+the application.
+
+In Spine, Projections are defined as Java classes, and their states are defined as
+Protobuf messages.
+
+### Repository
+
+Repository  is a mechanism for encapsulating storage, retrieval, and search behavior which emulates
```
Comment (armiol) [inline docs/introduction/concepts.md]: Please kill an extra space after `Repository`.

File: docs/introduction/concepts.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+* [`ProjectionRepository`](https://spine.io/core-java/javadoc/server/io/spine/server/projection/ProjectionRepository.html).
+
+### Snapshot
+
+Snapshot is a state of an Aggregate. A snapshot ”sits” in between events in the history of
+the Aggregate to make restoring faster.
+
+When an Aggregate is loaded, the `AggregateStorage` reads events backwards until encounters
+a snapshot. Then the snapshot is applied to the Aggregate, and trailing events are played to
+get the current state.
```
Comment (armiol) [inline docs/introduction/concepts.md]: I would rephrase as `...are played to obtain the most recent state.`

File: docs/introduction/concepts.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+There are two options for subscription:
+* receive changes of an Entity state for Projections, Process Managers and Aggregates; 
+* be notified of domain Events.
+
+## Architectural
+
+### Bounded Context
+
+Bounded Context is an autonomous component with its own domain model and its
+own Ubiquitous Language.Systems usually have multiple Bounded Contexts.
```
Comment (armiol) [inline docs/introduction/concepts.md]: Please add a space after `Language.`

File: docs/introduction/concepts.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+* be notified of domain Events.
+
+## Architectural
+
+### Bounded Context
+
+Bounded Context is an autonomous component with its own domain model and its
+own Ubiquitous Language.Systems usually have multiple Bounded Contexts. 
+
+For example, `Orders`, `UserManagement`, `Shipping` as examples of the contexts of an
```
Comment (armiol) [inline docs/introduction/concepts.md]: Let's kill the first "For example,":

```
`Orders`, `UserManagement`, `Shipping` are examples of the contexts of some online retail system.
```

File: docs/introduction/concepts.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+online retail system.
+
+### Message Buses
+
+#### Command Bus
+
+This is a message broker responsible for routing the command to its handler.
+Unlike a [Command Handler](#command-handler), it does not modify the application business model
```
Comment (armiol) [inline docs/introduction/concepts.md]: I would rephrase a bit: 
```It does not modify the application business model, nor produces domain events.```

File: docs/introduction/concepts.md (hunk 4 above)
Comment (alexander-yevsyukov) [reply docs/introduction/concepts.md]: I put it as the last item because:
  * I hope that the rule “The last is remembered” would work in this case.
  * The text that follows is based on the last item. Putting it first would break the logic flow.

File: docs/introduction/index.md [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+
+This exposes [`CommandService`](concepts.html#command-service), 
+[`QueryService`](concepts.html#query-service), and 
+[`SubscriptionService`](concepts.md#subscription-service) for client-side connections.
+
+## Repeating the cycle
+
+## Client application development
+ 
+
```
Comment (armiol) [inline docs/introduction/index.md]: Please kill the empty lines.

I would also append something like `This section is under development. For now, you may refer to the Javadoc of the Java client API or web client API.` — and put the links to the corresponding repositories (`core-java/client` and `web`).

File: docs/introduction/index.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+# Development Process
+
+Building a solution based on Spine Event Engine framework is an iterative process which consists
+of the stages described in this document.
+ 
+## Getting domain knowledge
+
+The purpose of this step is to find out what we're going to build and why.
+Consider using [EventStorming](https://eventstorming.com) or other DDD-based analysis 
+methodology for helping you grasp the domain knowledge from the experts.
```
Comment (armiol) [inline docs/introduction/index.md]: As I understand, you have omitted `Ubiquitous Language` to ease the way in. However, we should mention it someday. Let's discuss later.

File: docs/introduction/index.md [hunk 14]
```diff
@@ -218,7 +218,21 @@ This exposes [`CommandService`](concepts.html#command-service),
...
 
+The stages described above are repeated as another Bounded Context is added to the implementation,
+or as changes or extensions to the existing contexts are required. 
+ 
 ## Client application development
+
+Development of client applications may start after the [data types are defined](#defining-data-types).
+and Spine Model Compiler generates the code for all supported client platforms.
+ 
+Since entity classes like `Projection`s use composition with state types developers
```
Comment (armiol) [inline docs/introduction/index.md]: Please add a comma. `...with state types, developers ...`

File: docs/introduction/index.md [hunk 15]
```diff
@@ -218,7 +218,21 @@ This exposes [`CommandService`](concepts.html#command-service),
 
 ## Repeating the cycle
 
+The stages described above are repeated as another Bounded Context is added to the implementation,
+or as changes or extensions to the existing contexts are required. 
+ 
 ## Client application development
+
+Development of client applications may start after the [data types are defined](#defining-data-types).
+and Spine Model Compiler generates the code for all supported client platforms.
```
Comment (armiol) [inline docs/introduction/index.md]: Something is not right here. There is a period right before the start of this line.

File: docs/introduction/index.md (hunk 13 above)
Comment (alexander-yevsyukov) [reply docs/introduction/index.md]: Added issue #243 on this.
